feat: Gffread

tdayris opened the pull request:
<!-- Ensure that the PR title follows conventional commit style (<type>: <description>)-->
<!-- Possible types are here: https://github.com/commitizen/conventional-commit-types/blob/master/index.json -->

### Description

This PR adds [gffread](https://github.com/gpertea/gffread) to the list of available wrappers.

### QC
<!-- Make sure that you can tick the boxes below. -->

* [X] I confirm that:

For all wrappers added by this PR, 

* there is a test case which covers any introduced changes,
* `input:` and `output:` file paths in the resulting rule can be changed arbitrarily,
* either the wrapper can only use a single core, or the example rule contains a `threads: x` statement with `x` being a reasonable default,
* rule names in the test case are in [snake_case](https://en.wikipedia.org/wiki/Snake_case) and somehow tell what the rule is about or match the tools purpose or name (e.g., `map_reads` for a step that maps reads),
* all `environment.yaml` specifications follow [the respective best practices](https://stackoverflow.com/a/64594513/2352071),
* wherever possible, command line arguments are inferred and set automatically (e.g. based on file extensions in `input:` or `output:`),
* all fields of the example rules in the `Snakefile`s and their entries are explained via comments (`input:`/`output:`/`params:` etc.),
* `stderr` and/or `stdout` are logged correctly (`log:`), depending on the wrapped tool,
* temporary files are either written to a unique hidden folder in the working directory, or (better) stored where the Python function `tempfile.gettempdir()` points to (see [here](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir); this also means that using any Python `tempfile` default behavior works),
* the `meta.yaml` contains a link to the documentation of the respective tool or command,
* `Snakefile`s pass the linting (`snakemake --lint`),
* `Snakefile`s are formatted with [snakefmt](https://github.com/snakemake/snakefmt),
* Python wrapper scripts are formatted with [black](https://black.readthedocs.io).
* Conda environments use a minimal amount of channels, in recommended ordering. E.g. for bioconda, use (conda-forge, bioconda, nodefaults, as conda-forge should have highest priority and defaults channels are usually not needed because most packages are in conda-forge nowadays).

Commit:
[fix] (template): Missing code in wrappers' doc. Error #187

Commit:
Merge remote-tracking branch 'upstream/master'

tdayris committed:
Merge branch 'snakemake:master' into master

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Merge branch 'master' of https://github.com/tdayris/snakemake-wrappers

tdayris committed:
Merge remote-tracking branch 'upstream/master'

tdayris committed:
Merge remote-tracking branch 'upstream/master'

tdayris committed:
Merge remote-tracking branch 'upstreamhttp/master'

Commit:
Merge remote-tracking branch 'upstream/master'

johanneskoester committed:
perf: update utils/datavzrd/environment.yaml. (#566)

johanneskoester committed:
perf: update bio/bcftools/call/environment.yaml. (#567)

johanneskoester committed:
perf: update bio/bcftools/concat/environment.yaml. (#568)

johanneskoester committed:
perf: autobump bio/bcftools/index/environment.yaml (#570)

* perf: update bio/bcftools/index/environment.yaml.

* perf: update bio/bcftools/index/environment.yaml.

* perf: update bio/bcftools/index/environment.yaml.

github-actions[bot] committed:
chore: release 1.15.1 (#569)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

snakedeploy-bot[bot] committed:
perf: update bio/bgzip/environment.yaml. (#577)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

snakedeploy-bot[bot] committed:
perf: update bio/bellerophon/environment.yaml. (#576)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

snakedeploy-bot[bot] committed:
perf: update bio/assembly-stats/environment.yaml. (#575)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

snakedeploy-bot[bot] committed:
perf: update bio/adapterremoval/environment.yaml. (#573)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

johanneskoester committed:
chore: autobump conda envs (#571)

* Add autobump action

* fix paths

* dbg

* dbg branch

* add checkout

* dbg

* trigger rerun

* entity regex and add label

* dbg

* Update autobump.yml

* Update autobump.yml

snakedeploy-bot[bot] committed:
perf: update bio/delly/environment.yaml. (#584)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

snakedeploy-bot[bot] committed:
perf: update bio/busco/environment.yaml. (#581)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

snakedeploy-bot[bot] committed:
perf: update bio/clustalo/environment.yaml. (#582)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

snakedeploy-bot[bot] committed:
perf: update bio/fastp/environment.yaml. (#585)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

snakedeploy-bot[bot] committed:
perf: update bio/fastq_screen/environment.yaml. (#586)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

snakedeploy-bot[bot] committed:
perf: update bio/fasttree/environment.yaml. (#588)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

snakedeploy-bot[bot] committed:
perf: update bio/fastqc/environment.yaml. (#587)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

snakedeploy-bot[bot] committed:
perf: update bio/filtlong/environment.yaml. (#589)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

snakedeploy-bot[bot] committed:
perf: update bio/freebayes/environment.yaml. (#590)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

snakedeploy-bot[bot] committed:
perf: update bio/genefuse/environment.yaml. (#591)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

snakedeploy-bot[bot] committed:
perf: update bio/genomepy/environment.yaml. (#592)

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>

github-actions[bot] committed:
chore: release 1.15.2 (#578)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

FelixMoelder committed:
fix: set RG tag (#593)

johanneskoester committed:
Update conventional-prs.yml

snakedeploy-bot[bot] committed:
perf: autobump bio/deepvariant (#583)

Automatic update of bio/deepvariant.

Co-authored-by: snakedeploy-bot[bot] <115615832+snakedeploy-bot[bot]@users.noreply.github.com>
Co-authored-by: Johannes Köster <johannes.koester@uni-due.de>

christopher-schroeder committed:
feat: bazam wrapper (#580)

<!-- Ensure that the PR title follows conventional commit style (<type>:
<description>)-->
<!-- Possible types are here:
https://github.com/commitizen/conventional-commit-types/blob/master/index.json
-->

### Description

<!-- Add a description of your PR here-->

### QC
<!-- Make sure that you can tick the boxes below. -->

* [x] I confirm that:

For all wrappers added by this PR, 

* there is a test case which covers any introduced changes,
* `input:` and `output:` file paths in the resulting rule can be changed
arbitrarily,
* either the wrapper can only use a single core, or the example rule
contains a `threads: x` statement with `x` being a reasonable default,
* rule names in the test case are in
[snake_case](https://en.wikipedia.org/wiki/Snake_case) and somehow tell
what the rule is about or match the tools purpose or name (e.g.,
`map_reads` for a step that maps reads),
* all `environment.yaml` specifications follow [the respective best
practices](https://stackoverflow.com/a/64594513/2352071),
* wherever possible, command line arguments are inferred and set
automatically (e.g. based on file extensions in `input:` or `output:`),
* all fields of the example rules in the `Snakefile`s and their entries
are explained via comments (`input:`/`output:`/`params:` etc.),
* `stderr` and/or `stdout` are logged correctly (`log:`), depending on
the wrapped tool,
* temporary files are either written to a unique hidden folder in the
working directory, or (better) stored where the Python function
`tempfile.gettempdir()` points to (see
[here](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir);
this also means that using any Python `tempfile` default behavior
works),
* the `meta.yaml` contains a link to the documentation of the respective
tool or command,
* `Snakefile`s pass the linting (`snakemake --lint`),
* `Snakefile`s are formatted with
[snakefmt](https://github.com/snakemake/snakefmt),
* Python wrapper scripts are formatted with
[black](https://black.readthedocs.io).
* Conda environments use a minimal amount of channels, in recommended
ordering. E.g. for bioconda, use (conda-forge, bioconda, nodefaults, as
conda-forge should have highest priority and defaults channels are
usually not needed because most packages are in conda-forge nowadays).

github-actions[bot] committed:
chore: release 1.16.0 (#596)

:robot: I have created a release \*beep\* \*boop\*
---
##
[1.16.0](https://www.github.com/snakemake/snakemake-wrappers/compare/v1.15.2...v1.16.0)
(2022-10-13)


### Features

* bazam wrapper
([#580](https://www.github.com/snakemake/snakemake-wrappers/issues/580))
([17e58e6](https://www.github.com/snakemake/snakemake-wrappers/commit/17e58e6f254ce429b3b76b841df29b573949e278))


### Bug Fixes

* set RG tag
([#593](https://www.github.com/snakemake/snakemake-wrappers/issues/593))
([506a083](https://www.github.com/snakemake/snakemake-wrappers/commit/506a08391b56d4b53edda2c6555b9b9d404d4f94))


### Performance Improvements

* autobump bio/deepvariant
([#583](https://www.github.com/snakemake/snakemake-wrappers/issues/583))
([9b7c4fe](https://www.github.com/snakemake/snakemake-wrappers/commit/9b7c4feec69d05fc7d6286dcdfdc65802cb93317))
---


This PR was generated with [Release
Please](https://github.com/googleapis/release-please). See
[documentation](https://github.com/googleapis/release-please#release-please).

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

tdayris committed:
Merge remote-tracking branch 'upstream/master'

tdayris committed:
Merge remote-tracking branch 'upstream/master'

tdayris committed:
Merge remote-tracking branch 'upstream/master'

tdayris committed:
revert

tdayris committed:
revert

tdayris committed:
Merge branch 'master' of https://github.com/tdayris/snakemake-wrappers

tdayris committed:
gffread wrapper

tdayris committed:
revert

tdayris committed:
black re-formatting

fgvieira commented on bio/gffread/environment.yaml:
```diff
@@ -0,0 +1,6 @@
+channels:
+  - bioconda
+  - conda-forge
```
```suggestion
  - conda-forge
  - bioconda
```

fgvieira commented on bio/gffread/test/Snakefile:
```diff
@@ -0,0 +1,20 @@
+    input:
+        fasta="genome.fasta",
+        annotation="annotation.gtf",
+        # ids="",  # Optional path to records to kkep
```
```suggestion
        # ids="",  # Optional path to records to keep
```

fgvieira commented on bio/gffread/wrapper.py:
```diff
@@ -0,0 +1,69 @@
+extra = snakemake.params.get("extra", "")
+log = snakemake.log_fmt_shell(stdout=False, stderr=True)
+
+annotation = snakemake.input["annotation"]
+records = snakemake.output["records"]
```
```suggestion
annotation = snakemake.input.annotation
records = snakemake.output.records
```

fgvieira commented on bio/gffread/environment.yaml:
```diff
@@ -0,0 +1,6 @@
+  - conda-forge
+  - nodefaults
+dependencies:
+  - gffread
```
```suggestion
  - gffread =0.12.7
```

fgvieira commented on bio/gffread/wrapper.py:
```diff
@@ -0,0 +1,69 @@
+
+
+# Output format control
+if records.endswith(".gtf") and (" -T " not in extra):
+    extra += " -T "
+elif (records.endswith(".bed")) and (" --bed " not in extra):
+    extra += " --bed "
+elif (records.endswith(".tlf")) and (" --tlf " not in extra):
+    extra += " --tlf "
```
Are the outputs mutually exclusive, or can it output (e.g.) both a `gtf` and a `bed`?

tdayris replied:
There is only one output file, the format may be changed with these flags

fgvieira commented on bio/gffread/wrapper.py:
```diff
@@ -0,0 +1,69 @@
+records = snakemake.output["records"]
+
+# Input format control
+if (annotation.endswith(".bed")) and (" --in-bed " not in extra):
```
The check `" --in-bed " not in extra` won't work if `extra` starts with `--in-bed` (`extra = "--in-bed some/file.bed"`).

Is this check necessary? You can just state in the `meta.yaml` that those options are automatically inferred.

tdayris committed:
Update bio/gffread/environment.yaml

Co-authored-by: Filipe G. Vieira <1151762+fgvieira@users.noreply.github.com>

tdayris committed:
Update bio/gffread/test/Snakefile

Co-authored-by: Filipe G. Vieira <1151762+fgvieira@users.noreply.github.com>

tdayris committed:
Update bio/gffread/wrapper.py

Co-authored-by: Filipe G. Vieira <1151762+fgvieira@users.noreply.github.com>

tdayris committed:
Update bio/gffread/environment.yaml

Co-authored-by: Filipe G. Vieira <1151762+fgvieira@users.noreply.github.com>

tdayris committed:
Update wrapper.py

Unknown format now raises an error, extra parameters are not checked any more. A note will be added in `meta.yaml`

tdayris committed:
fix format error

tdayris committed:
fix format detection

fgvieira commented on bio/gffread/test/Snakefile:
```diff
@@ -0,0 +1,20 @@
+    input:
+        fasta="genome.fasta",
+        annotation="annotation.gtf",
+        # ids="",  # Optional path to records to keep
+        # nids="",  # Optional path to records to drop
+        # seq_info="",  # Optional path to sequence information
+        # sort_by="",  # Optional path to the ordered list of reference sequences
+        # attr="",  # Optional annotation attributes to keep.
+        # chr_replace="",  # Optional path to <original_ref_ID> <new_ref_ID>
```
Not sure if it is feasible but, since there seems to be some many extra inputs, it would be nice to also have tests for some of them. Is it too much of a hassle?

tdayris committed:
raise error on incompatible input files

tdayris committed:
Merge branch 'master' into gffread
